### PR TITLE
[SPARK-6996][SQL] Support map types in java beans

### PR DESCRIPTION
**Pull request by punya**
@liancheng @mengxr this is similar to #5146.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -1260,18 +1273,41 @@ class SQLContext(@transient val sparkContext: SparkContext)
       case c: Class[_] if c == classOf[java.sql.Date] => (DateType, true)
       case c: Class[_] if c == classOf[java.sql.Timestamp] => (TimestampType, true)
 
-      case c: Class[_] if c.isArray =>
-        val (dataType, nullable) = inferDataType(c.getComponentType)
+      case _ if typeToken.isArray =>
+        val (dataType, nullable) = inferDataType(typeToken.getComponentType)
+        (ArrayType(dataType, nullable), true)
+
+      case _ if iterableType.isAssignableFrom(typeToken) => {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Spark style avoids `{ }` for `case` statements.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 2]
```diff
@@ -18,13 +18,15 @@
...
+import java.lang.{Iterable => JIterable}
+import java.util.{Iterator => JIterator, Map => JMap, Properties}
 
 import scala.collection.JavaConversions._
 import scala.collection.immutable
-import scala.language.implicitConversions
+import scala.language.{existentials, implicitConversions}
 import scala.reflect.runtime.universe.TypeTag
 
+import com.google.common.reflect.TypeToken
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Nit: blank line to separate imports

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 3]
```diff
@@ -1222,20 +1224,31 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
-    val (dataType, _) = inferDataType(beanClass)
+    val (dataType, _) = SQLContext.inferDataType(TypeToken.of(beanClass))
     dataType.asInstanceOf[StructType].fields.map { f =>
       AttributeReference(f.name, f.dataType, f.nullable)()
     }
   }
 
+}
+
+object SQLContext {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: As long as we are going to break this out into its own object, can we move it to its own file and maybe name it something like `JavaTypeInference`?

**Comment by marmbrus** [comment]
Thanks for working on this.  I did not know this was possible given erasure!  A few minor comments, otherwise LGTM.

**Comment by punya** [comment]
@marmbrus relative to my earlier commit, I removed the code to handle Java `iterable`s as `ArrayType` because I couldn't get it to work at runtime. I think it's Catalyst-related and my knowledge isn't deep enough to figure out what's going on.

**Comment by marmbrus** [comment]
Have you looked in CatalystConverters?  There is handler for JMap, but
maybe not for JIterable?

**Comment by punya** [comment]
Thanks, `CatalystTypeConverters` was exactly what I needed to look at!

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

**Pull request closed without merging**